Opentrons/opentrons
# refactor(app): Add bare-minimum client calls for BR update

## overview

~Currently blocked by #3765, PR to be rebased once it is merged.~ Closes #3359 and #3361

This PR continues on the work of 3765 to enable the full BR migration/update flow from dispatching a single action. We'll still need more selectors and UI to actually dispatch said action, but this should be the bulk of the plumbing. Please read #3765 for technical background and decision making 

## changelog

- refactor(app): Add bare-minimum client calls for BR update

## review requests

The flow that this PR implements (including what was implemented in 3765) is:

![br-update-flow](https://user-images.githubusercontent.com/2963448/61718366-d714a280-ad30-11e9-9591-f15b2686c556.png)

**If you're going to test this, it will take a robot fully from Balena to buildroot**. If possible, you may want to do this on a freshly-minted Balena RPi with `ENABLE_VIRTUAL_SMOOTHIE=true`. Otherwise, code reviews are very much appreciated if you don't have the time to functionally test this.

We also don't currently have BR builds up for 3.10.1, so you will probably have to spoof your `app-shell/package.json` down to `3.10.0`.

### real-life testing instructions

```shell
make -C app dev OT_APP_DEV_INTERNAL__ENABLE_BUILD_ROOT=1
```

Dispatch a `buildroot:START_UPDATE` action to a robot name, e.g.:

```js
store.dispatch({type: 'buildroot:START_UPDATE', payload: 'opentrons-OT2P20180418A06'})
```

**Pro-tip**: Use the Chrome console rather than the Redux devtools to dispatch the action so you have up-arrow history as you're testing. Copy-paste in the app's devtools is very broken at least on macOS for some reason

Then, keep an eye on the Redux devtools for the following actions depending on which path in the flow you're following (this section copied / pasted / expanded on from 3765):

- `buildroot:START_PREMIGRATION`, `buildroot:PREMIGRATION_STARTED`, `buildroot:PREMIGRATION_DONE`
    - Robot was not ready for migration, so the update server was bootstrapped
- `robotApi:START_UPDATE`
    - Start update request will be re-dispatched after bootstrap is done and robot is migration capable
- `robotApi:[REQUEST|RESPONSE|ERROR]__POST__/session/update[/migration]/begin`
    - Token request, response, or error
    - Request path will change based on migration of BR update
- `robotApi:[REQUEST|RESPONSE|ERROR]__POST__/session/update[/migration]/cancel`
    - Token cancel request if initial token request came back 409
    - Request path will change based on migration of BR update
- `robotApi:[REQUEST|RESPONSE|ERROR]__POST__/session/update[/migration]/:token/status`
    - Status check poll on a 2s interval
    - **Please let me know if you can cause this interval to keep polling when it shouldn't be**
- `buildroot:UPLOAD_FILE`
    - File upload request made from app-shell side to avoid buffering the whole update file into memory
- `robotApi:[REQUEST|RESPONSE|ERROR]__POST__/session/update[/migration]/:token/commit`
    - Commit request once the file has been uploaded + verified by BR
- `robotApi:[REQUEST|RESPONSE|ERROR]__POST__/session[/update]/restart`
    - Restart request once file has been committed
    - Path is read from `capabilities` because varies based on BR migration vs update
- `buildroot:UNEXPECTED_ERROR`
    - This action needs fleshing out but punted to reduce scope
    - Something unexpected happened. Potential causes:
        - No BR file downloaded (check your `app-shell/package.json` in case you need to spoof it)
        - Update server bootstrap failed (check robot logs)
        - BR file upload request failed (check robot logs)
